src/collect/mod.rs

```diff
@@ -0,0 +1,252 @@
+//! Trace collectors.
+
+use crate::{errors::HWTracerError, Trace};
+use core::arch::x86_64::__cpuid_count;
+use libc::size_t;
+use strum::IntoEnumIterator;
+use strum_macros::EnumIter;
+
+#[cfg(collector_perf)]
+pub(crate) mod perf;
+#[cfg(collector_perf)]
+pub(crate) use perf::PerfTraceCollector;
+
+const PERF_DFLT_DATA_BUFSIZE: size_t = 64;
+const PERF_DFLT_AUX_BUFSIZE: size_t = 1024;
+const PERF_DFLT_INITIAL_TRACE_BUFSIZE: size_t = 1024 * 1024; // 1MiB
+
+/// The interface offered by all trace collectors.
+pub trait TraceCollector: Send + Sync {
+    /// Return a `ThreadTraceCollector` for tracing the current thread.
+    fn thread_collector(&self) -> Box<dyn ThreadTraceCollector>;
+}
+
+pub trait ThreadTraceCollector {
+    /// Start recording a trace.
+    ///
+    /// Tracing continues until [stop_collector](trait.ThreadTraceCollector.html#method.stop_collector) is called.
```

Comment:
I *think* you can just say "Tracing continues until [stop_collector] is called." and rustdoc will do the right thing. I might be wrong!

Reply:
The entire crate's docstrings are in dire need of attention. I was going to do a follow up PR.

Reply:
OK fair enough.